File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.cpp
class RandomizedSet {
public:
    vector<int> buffer;
    unordered_map<int, int> mp;

    bool find(int& val) {
        return mp.contains(val);
    }

    bool insert(int val) {
        if(find(val))
            return false;
        buffer.push_back(val);
        mp[val] = buffer.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        if(!find(val))
            return false;

        auto it = mp[val];
        swap(buffer[it], buffer[buffer.size() - 1]);
        mp[buffer[it]] = it;
        buffer.pop_back();
        mp.erase(val);
        return true;
    }
    
    int getRandom() {
        return buffer[rand() % buffer.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.go
type RandomizedSet struct {
    sl []int
    mp map[int]int
}


func Constructor() RandomizedSet {
    return RandomizedSet {
        sl: []int{},
        mp: map[int]int{},
    }
}


func (this *RandomizedSet) Insert(val int) bool {
    if this.Contains(val) {
        return false
    }
    this.sl = append(this.sl, val)
    this.mp[val] = len(this.sl) - 1
    return true
}


func (this *RandomizedSet) Remove(val int) bool {
    if !this.Contains(val) {
        return false
    }

    idx := this.mp[val]
    lastVal := this.sl[len(this.sl) - 1]
    this.sl[idx] = lastVal
    this.mp[lastVal] = idx
    this.sl = this.sl[:len(this.sl) - 1]
    delete(this.mp, val)
    return true
}


func (this *RandomizedSet) GetRandom() int {
    if len(this.mp) == 0 {
        return 0
    }
    return this.sl[rand.Intn(len(this.sl))]
}

func (this *RandomizedSet) Contains(val int) bool {
    _, ok := this.mp[val]
    return ok
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */